lib/Controller/PageController.php

```diff
@@ -125,14 +137,28 @@ public function index(): TemplateResponse {
 				$paths = array_map(function (Node $node) use ($userFolder) {
 					return substr(dirname($node->getPath()), strlen($userFolder->getPath()));
 				}, $search);
-				$this->nomediaPathsCache->set($key, $paths, 60 * 60 * 24 * 28);
+				$this->nomediaPathsCache->set($key, $paths, 60 * 60 * 24 * 28); // 28 days
 			}
 		} catch (InvalidPathException | NotFoundException | NotPermittedException | NoUserException $e) {
 			$this->logger->error($e->getMessage());
 		}
 
 		$this->initialState->provideInitialState('nomedia-paths', $paths);
 
+
+		$key = $user->getUID() . ':' . $userFolder->getEtag();
+		$tagCounts = $this->tagCountsCache->get($key);
+		if ($tagCounts === null) {
+			$tags = $this->tagManager->getAllTags(true);
+			$tagCounts = [];
+			foreach ($tags as $tag) {
+				$search = $userFolder->search(new SearchQuery(new SearchComparison(ISearchComparison::COMPARE_EQUAL, 'systemtag', $tag->getName()), 0, 0, [], $user));
+				$tagCounts[$tag->getName()] = count($search);
+			}
+			$this->tagCountsCache->set($key, $tagCounts, 60 * 60 * 24 * 7); // 7 days
```

Comment:
Shouldn't this be invalidated when the number of item changes ?
I am not familiar enough with `ICache`.

Reply:
Yeah, I'm not super confident about this bit. I found it important to add caching. Currently it will invalidate based on the user folder etag, but I'm not sure if that is affected by tags. Improvements welcome!

Reply:
How do you retrieve the list of files for a tag ? maybe we could invalidate there too?

Reply:
The client fetches files [here](https://github.com/nextcloud/photos/blob/master/src/services/TaggedImages.js). After the file list has been fetched for a tag on the client (which only happens if the tag is visible), the cached tagCount for that tag is overridden with the actual length of the file list. The point of this initial state (and cache) is to avoid having to wait for all tags' file lists to have been fetched until the top 10 tags can be displayed. Thus, if this cache entry is slightly outdated, it shouldn't matter too much, I think.

cc @skjnldsv What do you think?

Reply:
Sorry, I do not have the space for such a deep context switch :/

Reply:
Ah yes, it is a DAV request on an endpoint not controlled by photos.
Either listen to https://github.com/nextcloud/server/blob/ecad09b38a152957c771139562e61f7ecb1a6879/lib/public/SystemTag/MapperEvent.php#L37

Or simply recreate the cache every 30 min or so ? Is the search taking a long time ?

Reply:
If the search is fast, maybe we do not need to cache it.

Reply:
Listening to a tag event is probably the best approach.